File: packages/csharp/HtmlToMarkdown/Visitor/NodeType.cs
namespace HtmlToMarkdown.Visitor;

/// <summary>
/// Enumeration of all HTML node types recognized by the converter.
/// Used to classify elements during visitor callbacks.
/// </summary>
public enum NodeType
{
    // Text and Generic
    Text = 0,
    Element = 1,

    // Block Elements
    Heading = 2,
    Paragraph = 3,
    Div = 4,
    Blockquote = 5,
    Pre = 6,
    HorizontalRule = 7,

    // Lists
    List = 8,
    ListItem = 9,
    DefinitionList = 10,
    DefinitionTerm = 11,
    DefinitionDescription = 12,

    // Tables
    Table = 13,
    TableRow = 14,
    TableCell = 15,
    TableHeader = 16,
    TableBody = 17,
    TableHead = 18,
    TableFoot = 19,

    // Inline Formatting
    Link = 20,
    Image = 21,
    Strong = 22,
    Em = 23,
    Code = 24,
    Strikethrough = 25,
    Underline = 26,
    Subscript = 27,
    Superscript = 28,
    Mark = 29,
    Small = 30,
    LineBreak = 31,
    Span = 32,

    // Semantic HTML5
    Article = 33,
    Section = 34,
    Nav = 35,
    Aside = 36,
    Header = 37,
    Footer = 38,
    Main = 39,
    Figure = 40,
    FigCaption = 41,
    Time = 42,
    Details = 43,
    Summary = 44,

    // Forms
    Form = 45,
    Input = 46,
    Select = 47,
    Option = 48,
    Button = 49,
    TextArea = 50,
    Label = 51,
    FieldSet = 52,
    Legend = 53,

    // Media
    Audio = 54,
    Video = 55,
    Picture = 56,
    Source = 57,
    IFrame = 58,
    Svg = 59,
    Canvas = 60,

    // Advanced/Semantic
    Ruby = 61,
    Rt = 62,
    Rp = 63,
    Abbr = 64,
    Kbd = 65,
    Samp = 66,
    Var = 67,
    Cite = 68,
    Quote = 69,
    Del = 70,
    Ins = 71,
    Data = 72,
    Meter = 73,
    Progress = 74,
    Output = 75,
    Template = 76,
    Slot = 77,

    // Document Structure
    Html = 78,
    Head = 79,
    Body = 80,
    Title = 81,
    Meta = 82,
    LinkTag = 83,
    Style = 84,
    Script = 85,
    Base = 86,

    // Custom/Unknown
    Custom = 87
}

File: packages/java/src/main/java/io/github/goldziher/htmltomarkdown/visitor/NodeType.java
package io.github.goldziher.htmltomarkdown.visitor;

/**
 * HTML node type enumeration covering all element types supported by the
 * visitor pattern.
 *
 * <p>Maps directly to the C FFI NodeType enum. This enum categorizes HTML
 * elements for coarse-grained visitor dispatch.
 *
 * @since 2.17.0
 */
public enum NodeType {
    /** Text node. */
    TEXT,

    /** Generic element node. */
    ELEMENT,

    /** Heading element (h1-h6). */
    HEADING,

    /** Paragraph element. */
    PARAGRAPH,

    /** Division element. */
    DIV,

    /** Blockquote element. */
    BLOCKQUOTE,

    /** Preformatted text element. */
    PRE,

    /** Horizontal rule element. */
    HR,

    /** List element (ul, ol, dl). */
    LIST,

    /** List item element. */
    LIST_ITEM,

    /** Definition list element. */
    DEFINITION_LIST,

    /** Definition term element. */
    DEFINITION_TERM,

    /** Definition description element. */
    DEFINITION_DESCRIPTION,

    /** Table element. */
    TABLE,

    /** Table row element. */
    TABLE_ROW,

    /** Table cell element. */
    TABLE_CELL,

    /** Table header cell element. */
    TABLE_HEADER,

    /** Table body element. */
    TABLE_BODY,

    /** Table head element. */
    TABLE_HEAD,

    /** Table foot element. */
    TABLE_FOOT,

    /** Link element. */
    LINK,

    /** Image element. */
    IMAGE,

    /** Strong/bold element. */
    STRONG,

    /** Emphasis/italic element. */
    EM,

    /** Code element. */
    CODE,

    /** Strikethrough element. */
    STRIKETHROUGH,

    /** Underline element. */
    UNDERLINE,

    /** Subscript element. */
    SUBSCRIPT,

    /** Superscript element. */
    SUPERSCRIPT,

    /** Mark element. */
    MARK,

    /** Small element. */
    SMALL,

    /** Line break element. */
    BR,

    /** Span element. */
    SPAN,

    /** Article element. */
    ARTICLE,

    /** Section element. */
    SECTION,

    /** Navigation element. */
    NAV,

    /** Aside element. */
    ASIDE,

    /** Header element. */
    HEADER,

    /** Footer element. */
    FOOTER,

    /** Main element. */
    MAIN,

    /** Figure element. */
    FIGURE,

    /** Figure caption element. */
    FIGCAPTION,

    /** Time element. */
    TIME,

    /** Details element. */
    DETAILS,

    /** Summary element. */
    SUMMARY,

    /** Form element. */
    FORM,

    /** Input element. */
    INPUT,

    /** Select element. */
    SELECT,

    /** Option element. */
    OPTION,

    /** Button element. */
    BUTTON,

    /** Textarea element. */
    TEXTAREA,

    /** Label element. */
    LABEL,

    /** Fieldset element. */
    FIELDSET,

    /** Legend element. */
    LEGEND,

    /** Audio element. */
    AUDIO,

    /** Video element. */
    VIDEO,

    /** Picture element. */
    PICTURE,

    /** Source element. */
    SOURCE,

    /** Iframe element. */
    IFRAME,

    /** SVG element. */
    SVG,

    /** Canvas element. */
    CANVAS,

    /** Ruby element. */
    RUBY,

    /** Ruby text element. */
    RT,

    /** Ruby parenthesis element. */
    RP,

    /** Abbreviation element. */
    ABBR,

    /** Keyboard element. */
    KBD,

    /** Sample element. */
    SAMP,

    /** Variable element. */
    VAR,

    /** Citation element. */
    CITE,

    /** Quote element. */
    Q,

    /** Deleted element. */
    DEL,

    /** Inserted element. */
    INS,

    /** Data element. */
    DATA,

    /** Meter element. */
    METER,

    /** Progress element. */
    PROGRESS,

    /** Output element. */
    OUTPUT,

    /** Template element. */
    TEMPLATE,

    /** Slot element. */
    SLOT,

    /** HTML root element. */
    HTML,

    /** Head element. */
    HEAD,

    /** Body element. */
    BODY,

    /** Title element. */
    TITLE,

    /** Meta element. */
    META,

    /** Link tag element. */
    LINK_TAG,

    /** Style element. */
    STYLE,

    /** Script element. */
    SCRIPT,

    /** Base element. */
    BASE,

    /** Custom/unknown element. */
    CUSTOM;

    /**
     * Convert from C FFI node type value to Java enum.
     *
     * @param cValue the C FFI enum value
     * @return the corresponding Java NodeType
     * @throws IllegalArgumentException if the value is not recognized
     */
    public static NodeType fromCValue(final int cValue) {
        return values()[cValue];
    }

    /**
     * Convert to C FFI node type value.
     *
     * @return the C FFI enum value
     */
    public int toCValue() {
        return ordinal();
    }
}
